agentMET4FOF/metrological_streams.py

```diff
@@ -1,54 +1,87 @@
+    created: from dataset file (`set_data_source`), or a waveform generator function
+    (`set_generator_function`). Alternatively, override the `next_sample` function if
+    neither option suits the application. For generator functions, `sfreq` is a
+    required variable to be set on `init` which sets the sampling frequency and the
+    time-step which occurs when `next_sample()` is called.
     """
 
     def __init__(self):
         super().__init__()
 
-    @staticmethod
-    def _time():
-        return time.time()
+    def set_generator_function(self, generator_function=None, uncertainty_generator=None, sfreq=None, **kwargs):
+        """
+        Sets the a generator and uncertainty based on user-defined functions. By default,
+        this function resorts to a sine wave generator function and a constant (zero) uncertainty.
+        Initialisation of the generator's parameters should be done here such as setting the sampling frequency and
+        wave frequency. For setting it with a dataset instead, see `set_data_source`. Overwrites the default
+        set_generator_function method in DataStreamMET4FO
+
+        Parameters
+        ----------
+        generator_function : method
+            A generator function which takes in at least one argument `time` which
+            will be used in `next_sample`.
+
+        uncertainty_generator : method
+
+        sfreq : int
+            Sampling frequency.
 
-    @staticmethod
-    def _time_unc():
-        return time.get_clock_info("time").resolution
+        **kwargs
+            Any additional keyword arguments to be supplied to the generator function.
+            The ``**kwargs`` will be saved as `uncertainty_parameters`.
+            The generator function call for every sample will be supplied with the
+            ``**uncertainty_parameters``.
 
-    @staticmethod
-    def _value(timestamp):
-        return 1013.25 + 10 * np.sin(timestamp)
+        """
+        # Call the set_generator_function from the parent class to set the generator function
+        super().set_generator_function(generator_function=generator_function, sfreq=sfreq, **kwargs)
 
-    @staticmethod
-    def _value_unc():
-        return 0.5
+        self.uncertainty_parameters = kwargs
 
-    @property
-    def current_datapoint(self):
-        t = self._time()
-        ut = self._time_unc()
-        v = self._value(t)
-        uv = self._value_unc()
+        # resort to default wave generator if one is not supplied
+        if uncertainty_generator is None:
+            warnings.warn("No uncertainty generator function specified. Setting to default (zero).")
+            self.generator_function_unc = self.default_uncertainty_generator
+        else:
+            self.generator_function_unc = uncertainty_generator
+        return self.generator_function_unc
 
-        return np.array((t, ut, v, uv))
 
+    def default_uncertainty_generator(self, time):
```

Comment:
```suggestion
    def default_uncertainty_generator(self) -> Tuple[float, float]:
```
See next comment for required import.